effekt/jvm/src/test/scala/effekt/PolymorphismBoxingTests.scala

```diff
@@ -0,0 +1,286 @@
+package effekt
+
+import java.io.File
+import effekt.core.{Block, Definition, DirectApp, PolymorphismBoxing, Pure, Run, Stmt}
+import effekt.context.Context
+import effekt.source.{IdDef, Import, ModuleDecl}
+import kiama.{parsing, util}
+import effekt.symbols.{Module, Name, TypeConstructor, TypeSymbol, ValueSymbol, ValueType}
+import effekt.source
+import effekt.util.messages
+import effekt.util.messages.DebugMessaging
+import kiama.parsing.{Failure, NoSuccess, Success}
+import kiama.util.Severities
+
+abstract class AbstractPolymorphismBoxingTests extends munit.FunSuite {
+
+  def boxDef(tpe: ValueType.ValueTypeApp): List[symbols.Symbol] = {
+    val tpeCns: symbols.TypeConstructor.Record =
+      symbols.TypeConstructor.Record(symbols.LocalName("Boxed" ++ tpe.constructor.name.name), List(), null)
+    tpeCns.constructor =
+      symbols.Constructor(symbols.LocalName("MkBoxed" ++ tpe.constructor.name.name), List(), null, tpeCns)
+    tpeCns.constructor.fields = List(
+      symbols.Field(symbols.LocalName("unbox" ++ tpe.constructor.name.name),
+        symbols.ValueParam(symbols.LocalName("value"), Some(tpe)), tpeCns.constructor))
+    List(tpeCns, tpeCns.constructor, tpeCns.constructor.fields.head)
+  }
+
+  val boxtpes = symbols.builtins.rootTypes.values.flatMap {
+    case t: TypeConstructor.ExternType => boxDef(ValueType.ValueTypeApp(t, List()))
+    case _ => Nil
+  }.map { c => (c.name.name, c) }.toMap
+
+  /** Mock context for Polymorphism boxing.
+   * Only implements what is actually used by [[core.PolymorphismBoxing]]
+   */
+  object context extends Context(new util.Positions()) {
+    this.module = new Module(ModuleDecl("test", List(Import("effekt")), List()), util.StringSource("", "test")) {
+      override def findPrelude: Module = new Module(ModuleDecl("effekt", List(), List()), util.StringSource("", "effekt")) {
+        override def types: Map[String, TypeSymbol] = boxtpes.collect[String, symbols.TypeSymbol]{
+          case (k,t: symbols.TypeSymbol) => (k,t)
+        }
+      }
+    }
+
+    object messaging extends DebugMessaging
+
+    def contentsOf(path: String): Option[String] = None
+
+    def findSource(path: String): Option[kiama.util.Source] = None
+  }
+
+  val names = new core.Names(boxtpes ++
+    symbols.builtins.rootTypes ++ Map(
+    // TODO maybe add used names
+  ))
+
+  class ConsistentRenamer(names: core.Names, prefix: String = "l") extends core.Tree.Rewrite {
+    var bound: List[symbols.Symbol] = Nil
+    def withBindings[R](ids: List[symbols.Symbol])( f: => R ): R = {
+      val oldBound = bound
+      bound = ids ++ bound
+      val res = f
+      bound = oldBound
+      res
+    }
+    def withBinding[R](id: symbols.Symbol)( f: => R ): R = withBindings(List(id))(f)
+
+    override def pure: PartialFunction[Pure, Pure] = {
+      case Pure.ValueVar(id, tpe) => Pure.ValueVar(rewrite(id), tpe)
```

Comment:
Hmmmmm visiting the names seems like a very common case... Maybe we can change rewrite to also have a case for this?

Reply:
Agreed.
Thinking about it, having something like this for types, too, might even make the `PolymorphismBoxing` phase simpler 🤔

Reply:
Done.